lib/upload-manager.js

```diff
@@ -0,0 +1,287 @@
+	ReadableStream = require('stream').Readable,
+	crypto = require('crypto');
+
+// -----------------------------------------------------------------------------
+// Private
+// -----------------------------------------------------------------------------
+
+//@TODO: make into an option
+const RETRY_INTERVAL_MS = 1000;
+
+const DEFAULT_OPTIONS = {
+	parallelism: 4
+};
+
+class Chunk extends EventEmitter {
+
+	/**
+	 * Create a Chunk, representing a part of a file being uploaded
+	 * @param {BoxClient} client The Box SDK client
+	 * @param {string} sessionID The ID of the upload session the chunk belongs to
+	 * @param {?Buffer|string} chunk The chunk that was uploaded
+	 * @param {int} offset The bute offset within the file where this chunk begins
+	 * @param {int} totalSize The total size of the file this chunk belongs to
+	 */
+	constructor(client, sessionID, chunk, offset, totalSize) {
+
+		super();
+
+		this.client = client;
+		this.sessionID = sessionID;
+		this.chunk = chunk;
+		this.length = chunk.length;
+		this.offset = offset;
+		this.totalSize = totalSize;
+		this.data = null;
+		this.retry = null;
+		this.canceled = false;
+	}
+
+	/**
+	 * Get the final object representation of this chunk for the API
+	 * @returns {UploadPart} The chunk object
+	 */
+	getData() {
+
+		return this.data;
+	}
+
+	/**
+	 * Upload a chunk to the API
+	 * @returns {void}
+	 */
+	upload() {
+
+		this.client.files.uploadPart(this.sessionID, this.chunk, this.offset, this.totalSize, null, (err, data) => {
+
+			if (this.canceled) {
+				this.chunk = null;
+				return;
+			}
+
+			if (err) {
+				// handle the error or retry
+				if (err.statusCode) {
+					// an API error, probably not retryable!
+					this.emit('error', err);
+				} else {
+					// maybe a network error, retry
+					this.retry = setTimeout(() => this.upload(), RETRY_INTERVAL_MS);
+				}
+				return;
+			}
+
+			// Record the chunk data for commit, and try to free up the chunk buffer
+			this.data = data;
+			this.chunk = null;
+			this.emit('uploaded', data);
+		});
+	}
+
+	/**
+	 * Cancel trying to upload a chunk, preventing it from retrying and clearing
+	 * the associated buffer
+	 * @returns {void}
+	 */
+	cancel() {
+
+		clearTimeout(this.retry);
+		this.chunk = null;
+		this.canceled = true;
+	}
+}
+
+// -----------------------------------------------------------------------------
+// Public
+// -----------------------------------------------------------------------------
+
+//@TODO: rename to something better
+/** Manager for uploading a file in chunks */
+class UploadManager extends EventEmitter {
+
+	/**
+	 * Create an upload manager
+	 * @param {BoxClient} client The client to use to upload the file
+	 * @param {Object} uploadSessionInfo The upload session info to use for chunked upload
+	 * @param {ReadableStream|Buffer|string} file The file to upload
+	 * @param {int} size The size of the file to be uploaded
+	 * @param {Object} [options] Optional parameters
```

Comment:
Since `options` is the last param here, it's truly optional instead of merely nullable, so I marked it as such.